### DISCUSSION TITLE
A little confused about how to implement a ConsumerRebalanceListener

### DISCUSSION THREAD

**Discussion by jscheel** [aio-libs/aiokafka#1042] [Q&A]
Hi there, I'm a bit confused about using a ConsumerRebalanceListener to ensure I commit my offsets appropriately when manually managing my commits. Specifically, I'm looking at the warning here: https://aiokafka.readthedocs.io/en/stable/consumer.html#manual-vs-automatic-committing

If I have a batch consumer that matches the example given, how would I implement this?

```python
consumer = AIOKafkaConsumer(
    "my_topic",
    bootstrap_servers='localhost:9092',
    group_id="my_group",           # Consumer must be in a group to commit
    enable_auto_commit=False,      # Will disable autocommit
    auto_offset_reset="earliest",  # If committed offset not found, start
                                   # from beginning
)
await consumer.start()

batch = []
async for msg in consumer:
    batch.append(msg)
    if len(batch) == 100:
        await process_msg_batch(batch)
        await consumer.commit()
        batch = []
```

Would I use an asyncio lock to prevent `on_partitions_revoked` and `on_partitions_assigned` from returning until I call the next commit?

**Comment by vmaurin** [comment]
Hi, I had a similar question here https://github.com/aio-libs/aiokafka/discussions/935

We ended up implementing the pattern in the second snippets

```python
while True:
    records = consumer.getmany(...)
    offsets = {}
    for tp, messages in records.items():
        offsets[tp] = messages[-1].offset + 1
    await some_processing(records)
    try:
        consumer.commit(offsets)
    except (CommitFailedError, IllegalStateError) as err:
        logger.warning("Commit failed due to rebalancing, circle back to consume new messages",  exc_info=err)
```

**Comment by vmaurin** [comment]
And around `ConsumerRebalanceListener`, as it runs on a separated coroutine, it is easier to fall into a lot of trap, like any kind of await could resume your main consumer coroutine and you have to deal with a lot of concurrency. We do use it in my company, but mostly for cleanups

**Comment by marcosschroh** [comment]
Hi all. I have created a `micro fremework` based on `aiokafka` called [kstreams](https://kpn.github.io/kstreams/), which deals with this [common uses[(https://kpn.github.io/kstreams/stream/#rebalance-listener) cases. You can give it a try if you want. We are using a our solution to deal with `RebalanceListener` and `commits` works mostly all the time.
